### PR TITLE
FIX: f3d --help invalid output #1622

### PR DESCRIPTION
**Pull request by Schwarzemann**
I replaced `std::string_view` with `std::string` to format the command and the descriptions. I also used `std::setw()` for alignment. This will avoid strings going out of view or getting corrupted. Also each command for the mentioned section is now logged as strings using `f3d::log::info`.

Screenshot 2024-09-16 045032

Fix for: https://github.com/f3d-app/f3d/issues/1622

### CODE REVIEW

File: application/F3DOptionsTools.cxx [hunk 1]
```diff
@@ -225,9 +225,10 @@ void PrintHelp(const std::string& execName, const cxxopts::Options& cxxOptions)
   }
 
   f3d::log::info("\nExamples:");
-  for (const auto& [cmd, desc] : examples)
+  for (const auto& example : examples)
   {
-    ::PrintHelpPair(cmd, desc, 50);
```
Comment (mwestphal) [inline application/F3DOptionsTools.cxx:230]: this should still work afaics

Comment (mwestphal) [reply application/F3DOptionsTools.cxx:230]: yes, you can just use this method instead of printing manually

Comment (Schwarzemann) [reply application/F3DOptionsTools.cxx:230]: I tried it but it gets broken again. I am attaching a screenshot.
Screenshot 2024-09-16 093151

Comment (mwestphal) [reply application/F3DOptionsTools.cxx:230]: you could add a `separator` arg to PrintHelpPair at the end that default to `''`  and set it to `\n` in your case ?

**Comment by mwestphal** [comment]
by the way @Schwarzemann , we have  a discord, do not hesitate to join :) 

https://discord.f3d.app

File: application/F3DOptionsTools.cxx (hunk 1 above)
Comment (Schwarzemann) [reply application/F3DOptionsTools.cxx]: That seems to do the trick! Thanks!

**Comment by Schwarzemann** [comment]
> by the way @Schwarzemann , we have a discord, do not hesitate to join :)
> 
> https://discord.f3d.app

I definitely will!

File: application/F3DOptionsTools.cxx [hunk 2]
```diff
@@ -192,16 +192,16 @@ void PrintHelpPair(
   std::string_view key, std::string_view help, int keyWidth = 10, int helpWidth = 70)
 {
   std::stringstream ss;
-  ss << "  " << std::left << std::setw(keyWidth) << key << " " << std::setw(helpWidth) << help;
+  ss << "  " << std::left << std::setw(keyWidth) << key << " \n" << std::setw(helpWidth) << help;
```
Comment (mwestphal) [inline application/F3DOptionsTools.cxx:195]: this is not what I meant. I meanty to add a new argments to PrintHelpPair so this char here can be controlled.

**Review by mwestphal — CHANGES_REQUESTED** [review]
need some changes

File: application/F3DOptionsTools.cxx (hunk 2 above)
Comment (snoyer) [reply application/F3DOptionsTools.cxx:195]: or add the `\n` automatically if the key is longer that the padding maybe?
```cpp
  ss << "  " << std::left << std::setw(keyWidth) << key;
  if(key.size() > keyWidth){
   ss << "\n  " << std::setw(keyWidth) << " ";
  }
  ss << " " << std::setw(helpWidth) << help;
```
```
Examples:
  f3d file.vtu -xtgans                               View a unstructured mesh in a typical nice looking sciviz style       
  f3d file.glb -tuqap --hdri-file=file.hdr --hdri-ambient --hdri-skybox
                                                     View a gltf file in a realistic environment                           
  f3d file.ply -so --point-size=0 --comp=-2          View a point cloud file with direct scalars rendering                 
  f3d folder                                         View all files in folder       
```

File: application/F3DOptionsTools.cxx [hunk 3]
```diff
@@ -188,50 +188,71 @@ std::string CollapseName(const std::string_view& longName, const std::string_vie
...
+        std::stringstream descStream(desc);
+        std::string line;
+        while (std::getline(descStream, line))
+        {
+            f3d::log::info("   " + line);
+        }
+    }
+    f3d::log::info("\nReport bugs to https://github.com/f3d-app/f3d/issues");
+    f3d::log::setUseColoring(true);
+    f3d::log::waitForUser();
```
Comment (mwestphal) [inline application/F3DOptionsTools.cxx:255]: looks like uncorrect formatting changes

Comment (Schwarzemann) [reply application/F3DOptionsTools.cxx]: I restored the original format.

File: application/F3DOptionsTools.cxx [hunk 4]
```diff
@@ -188,50 +188,71 @@ std::string CollapseName(const std::string_view& longName, const std::string_vie
...
+    for (const ::CLIGroup& optionGroup : ::CLIOptions)
+    {
+        orderedCLIGroupNames.emplace_back(optionGroup.GroupName);
+    }
+    f3d::log::info(cxxOptions.help(orderedCLIGroupNames));
+    f3d::log::info("Keys:");
+    for (const auto& [key, desc] : f3d::interactor::getDefaultInteractionsInfo())
+    {
+        ::PrintHelpPair(key, desc);
+    }
```
[the remark this answers is not in the document]
Comment (Schwarzemann) [reply application/F3DOptionsTools.cxx]: I restored the original loop

File: application/F3DOptionsTools.cxx (hunk 2 above)
Comment (Schwarzemann) [reply application/F3DOptionsTools.cxx]: It should be fine rn.

**Review by Schwarzemann — COMMENTED** [review]
I made some changes in the last PR update.

File: application/F3DOptionsTools.cxx [hunk 5]
```diff
@@ -190,69 +190,63 @@ std::string CollapseName(const std::string_view& longName, const std::string_vie
...
+  const std::array<std::pair<std::string, std::string>, 4> examples = {{
     { execName + " file.vtu -xtgans",
       "View an unstructured mesh in a typical nice-looking sciviz style" },
     { execName + " file.glb -tuqap --hdri-file=file.hdr --hdri-ambient --hdri-skybox",
       "View a gltf file in a realistic environment" },
     { execName + " file.ply -so --point-size=0 --comp=-2",
       "View a point cloud file with direct scalars rendering" },
     { execName + " folder", "View all files in folder" },
-    }};
+  }};
```
Comment (mwestphal) [inline application/F3DOptionsTools.cxx]: incorrect formating changes

File: application/F3DOptionsTools.cxx [hunk 6]
```diff
@@ -188,20 +188,28 @@ std::string CollapseName(const std::string_view& longName, const std::string_vie
...
+  if (key.find(' ') != std::string::npos)
+  {
+    ss << key << "\n";
+    ss << std::setw(keyWidth) << " " << help;
+  }
+  else
+  {
+    ss << std::left << std::setw(keyWidth) << key;
+    ss << " " << help;
+  }
```
Comment (mwestphal) [inline application/F3DOptionsTools.cxx:203]: this is not what @snoyer suggested here: https://github.com/f3d-app/f3d/pull/1626#discussion_r1760686326

Lets implement what was suggested.

File: application/F3DOptionsTools.cxx [hunk 7]
```diff
@@ -214,21 +222,23 @@ void PrintHelp(const std::string& execName, const cxxopts::Options& cxxOptions)
...
     orderedCLIGroupNames.emplace_back(optionGroup.GroupName);
   }
+  
   f3d::log::info(cxxOptions.help(orderedCLIGroupNames));
+  
   f3d::log::info("Keys:");
   for (const auto& [key, desc] : f3d::interactor::getDefaultInteractionsInfo())
   {
-    ::PrintHelpPair(key, desc);
+    ::PrintHelpPair(key, desc, 20);
```
Comment (mwestphal) [inline application/F3DOptionsTools.cxx:233]: why changing this ?

File: application/F3DOptionsTools.cxx [hunk 8]
```diff
@@ -214,21 +222,23 @@ void PrintHelp(const std::string& execName, const cxxopts::Options& cxxOptions)
...
   }
 
   f3d::log::info("\nExamples:");
   for (const auto& [cmd, desc] : examples)
   {
-    ::PrintHelpPair(cmd, desc, 50);
+    ::PrintHelpPair(cmd, desc, 5);
```
Comment (mwestphal) [inline application/F3DOptionsTools.cxx:239]: why changing this ?

Comment (Schwarzemann) [reply application/F3DOptionsTools.cxx:239]: It makes the text more close to each other.

File: application/F3DOptionsTools.cxx (hunk 7 above)
Comment (Schwarzemann) [reply application/F3DOptionsTools.cxx:233]: The keys section looks more close to each other.

Comment (mwestphal) [reply application/F3DOptionsTools.cxx:233]: can you share sceenshot of the result ?

Comment (Schwarzemann) [reply application/F3DOptionsTools.cxx:233]: Yes most certainly. Here you go.


When I increase the number the gap between texts grow wider.

Comment (mwestphal) [reply application/F3DOptionsTools.cxx:233]: I dont see how this looks better than current version.

Comment (Schwarzemann) [reply application/F3DOptionsTools.cxx:233]: Okay I won't tamper with the looks then I will implement the solution that you mentioned earlier. Thanks for putting up.

File: application/F3DOptionsTools.cxx [hunk 9]
```diff
@@ -192,14 +192,19 @@ void PrintHelpPair(
   std::string_view key, std::string_view help, int keyWidth = 10, int helpWidth = 70)
 {
   std::stringstream ss;
-  ss << "  " << std::left << std::setw(keyWidth) << key << " " << std::setw(helpWidth) << help;
+  ss << "  " << std::left << std::setw(keyWidth) << key;
+  if (key.size() > static_cast<size_t>(keyWidth))
+  {
+    ss << "\n  " << std::setw(keyWidth) << " "; // Indent the second line
```
Comment (mwestphal) [inline application/F3DOptionsTools.cxx:198]: i dont think you need the keyWidth here thoug

Comment (Schwarzemann) [reply application/F3DOptionsTools.cxx]: Sorry you are right. I added it with an instinct but it's not necessary to check the second line.

**Review by mwestphal — APPROVED** [review]
@snoyer please review

File: application/F3DOptionsTools.cxx [hunk 10]
```diff
@@ -192,14 +192,19 @@ void PrintHelpPair(
   std::string_view key, std::string_view help, int keyWidth = 10, int helpWidth = 70)
 {
   std::stringstream ss;
-  ss << "  " << std::left << std::setw(keyWidth) << key << " " << std::setw(helpWidth) << help;
+  ss << "  " << std::left << std::setw(keyWidth) << key;
+  if (key.size() > static_cast<size_t>(keyWidth))
+  {
+    ss << "\n  ";
```
Comment (snoyer) [inline application/F3DOptionsTools.cxx:198]: @mwestphal do we want the `help` string to still be aligned? if so:
```suggestion
    ss << "\n  " << std::setw(keyWidth) << " ";
```
as per PR:
```
Examples:
  f3d file.vtu -xtgans                               View a unstructured mesh in a typical nice looking sciviz style       
  f3d file.glb -tuqap --hdri-file=file.hdr --hdri-ambient --hdri-skybox
   View a gltf file in a realistic environment                           
  f3d file.ply -so --point-size=0 --comp=-2          View a point cloud file with direct scalars rendering                 
  f3d folder                                         View all files in folder   
```
with suggestion:
```
Examples:
  f3d file.vtu -xtgans                               View a unstructured mesh in a typical nice looking sciviz style       
  f3d file.glb -tuqap --hdri-file=file.hdr --hdri-ambient --hdri-skybox
                                                     View a gltf file in a realistic environment                           
  f3d file.ply -so --point-size=0 --comp=-2          View a point cloud file with direct scalars rendering                 
  f3d folder                                         View all files in folder 
```

Comment (mwestphal) [reply application/F3DOptionsTools.cxx:198]: indeed, you are right @snoyer

**Review by mwestphal — CHANGES_REQUESTED** [review]
pleasse follow @snoyer recommendation

**Comment by Schwarzemann** [comment]
> pleasse follow @snoyer recommendation
 I applied the recommended change.

File: application/F3DOptionsTools.cxx [hunk 11]
```diff
@@ -214,11 +219,10 @@ void PrintHelp(const std::string& execName, const cxxopts::Options& cxxOptions)
   orderedCLIGroupNames.reserve(::CLIOptions.size());
   for (const ::CLIGroup& optionGroup : ::CLIOptions)
   {
-    // This ensure help is provided in the expected group order
```
Comment (mwestphal) [inline application/F3DOptionsTools.cxx:217]: please keep this comment

**Review by mwestphal — CHANGES_REQUESTED** [review]
small changes needed

File: application/F3DOptionsTools.cxx (hunk 11 above)
Comment (Schwarzemann) [reply application/F3DOptionsTools.cxx]: Done

File: application/F3DOptionsTools.cxx [hunk 12]
```diff
@@ -214,11 +219,10 @@ void PrintHelp(const std::string& execName, const cxxopts::Options& cxxOptions)
...
     orderedCLIGroupNames.emplace_back(optionGroup.GroupName);
   }
   f3d::log::info(cxxOptions.help(orderedCLIGroupNames));
-  f3d::log::info("Keys:");
+    f3d::log::info("Keys:");
```
[the remark this answers is not in the document]
Comment (Schwarzemann) [reply application/F3DOptionsTools.cxx]: Done

**Review by mwestphal — APPROVED** [review]

**Pull request merged**